src/Processors/QueryPlan/Optimizations/mergeFilterIntoJoinCondition.cpp

```diff
@@ -0,0 +1,361 @@
+            auto lhs_side = getExpressionSide(lhs, left_stream_allowed_nodes, right_stream_allowed_nodes);
+            auto rhs_side = getExpressionSide(rhs, left_stream_allowed_nodes, right_stream_allowed_nodes);
+
+            if (lhs_side == ExpressionSide::LEFT && rhs_side == ExpressionSide::RIGHT)
+            {
+                result.emplace_back(createConditionPart(lhs, rhs));
+                conjuncts_to_replace.insert(conjunct);
+                continue;
+            }
+            else if (rhs_side == ExpressionSide::LEFT && lhs_side == ExpressionSide::RIGHT)
+            {
+                result.emplace_back(createConditionPart(rhs, lhs));
+                conjuncts_to_replace.insert(conjunct);
+                continue;
+            }
+        }
+        rejected_conjuncts.push_back(conjunct);
+    }
+
+    bool trivial_filter = rejected_conjuncts.empty();
+    if (!result.empty())
+    {
+        /// There's a non-empty list of extracted condition parts.
+        /// After JOIN step these equalities will always evaluate to true.
+        for (const auto * & output : filter_dag.getOutputs())
+        {
+            auto it = conjuncts_to_replace.find(output);
+            if (it != conjuncts_to_replace.end())
+            {
+                output = &filter_dag.addColumn(ColumnWithTypeAndName(
+                    output->result_type->createColumnConst(1, 1),
+                    output->result_type,
+                    output->result_name));
+            }
+        }
+
+        if (rejected_conjuncts.size() == 1)
+        {
+            filter_dag.addOrReplaceInOutputs(filter_dag.addAlias(*rejected_conjuncts[0], filter_name));
+        }
+        else if (rejected_conjuncts.size() > 1)
+        {
+            FunctionOverloadResolverPtr func_builder_and = std::make_unique<FunctionToOverloadResolverAdaptor>(std::make_shared<FunctionAnd>());
+            filter_dag.addOrReplaceInOutputs(filter_dag.addFunction(func_builder_and, std::move(rejected_conjuncts), filter_name));
+        }
+        filter_dag.removeUnusedActions(/*allow_remove_inputs=*/false);
+    }
+
+    return { std::move(result), trivial_filter };
+}
+
+}
+
+size_t tryMergeFilterIntoJoinCondition(QueryPlan::Node * parent_node, QueryPlan::Nodes &  /*nodes*/, const Optimization::ExtraSettings &)
+{
+    if (parent_node->children.size() != 1)
+        return 0;
+
+    QueryPlan::Node * child_node = parent_node->children.front();
+
+    auto & parent = parent_node->step;
+    auto & child = child_node->step;
+
+    auto * filter_step = typeid_cast<FilterStep *>(parent.get());
+    auto * join_step = typeid_cast<JoinStepLogical *>(child.get());
+
+    if (!filter_step || !join_step)
+        return 0;
+
+    const auto & join_expressions = join_step->getExpressionActions();
+    auto & join_info = join_step->getJoinInfo();
+
+    if (join_info.kind == JoinKind::Full || join_info.kind == JoinKind::Left || join_info.kind == JoinKind::Right)
```

Comment:
I prefer to allow only for CROSS and INNER
Who knows maybe we can add another kind later :)